### PR TITLE
Azure.Provisioning: Move name validation to Infrastructure and expose for Aspire

### PR DESCRIPTION
**Pull request by tg-msft**

### CODE REVIEW

**Review by jsquire — APPROVED** [review]

File: sdk/provisioning/Azure.Provisioning/src/Infrastructure.cs [hunk 1]
```diff
@@ -93,6 +94,97 @@ public virtual void Remove(Provisionable resource)
...
+    /// </summary>
+    /// <param name="identifierName">The proposed identifier name.</param>
+    /// <returns>Whether the name is a valid bicep identifier name.</returns>
+    public static bool IsValidIdentifierName(string? identifierName)
+    {
+        if (string.IsNullOrEmpty(identifierName)) { return false; }
+        if (char.IsDigit(identifierName![0])) { return false; }
+        foreach (char ch in identifierName)
+        {
+            if (!char.IsLetterOrDigit(ch) && ch != '_')
```
Comment (eerhardt) [inline sdk/provisioning/Azure.Provisioning/src/Infrastructure.cs:109]: Are we concerned about non-ascii letters?

File: sdk/provisioning/Azure.Provisioning/src/Infrastructure.cs [hunk 2]
```diff
@@ -93,6 +94,97 @@ public virtual void Remove(Provisionable resource)
...
+            // TODO: This may be relaxed in the future to generate an automatic
+            // name rather than throwing
+            throw new ArgumentNullException(nameof(identifierName), $"{nameof(identifierName)} cannot be null.");
+        }
+        else if (identifierName.Length == 0)
+        {
+            throw new ArgumentException($"{nameof(identifierName)} cannot be empty.", nameof(identifierName));
+        }
+
+        StringBuilder builder = new(identifierName.Length);
```
Comment (eerhardt) [inline sdk/provisioning/Azure.Provisioning/src/Infrastructure.cs:169]: Probably doesn't matter too much, but if `identifierName` is already valid, why allocate a StringBuilder and a new string?

File: sdk/provisioning/Azure.Provisioning/src/Infrastructure.cs [hunk 3]
```diff
@@ -93,6 +94,97 @@ public virtual void Remove(Provisionable resource)
...
+    }
+
+    /// <summary>
+    /// Validates whether a given bicep identifier name is correctly formed of
+    /// letters, numbers, and underscores.
+    /// </summary>
+    /// <param name="identifierName">The proposed bicep identifier name.</param>
+    /// <exception cref="ArgumentNullException">Throws if null.</exception>
+    /// <exception cref="ArgumentException">Throws if empty or invalid.</exception>
+    public static void ValidateIdentifierName(string? identifierName)
```
Comment (eerhardt) [inline sdk/provisioning/Azure.Provisioning/src/Infrastructure.cs:124]: Would it be possible to take a `string? parameterName = null` parameter? That way this method could be used by other callers and they can just let the ArgumentException through.

For example:

```C#
    public static IResourceBuilder<T> WithParameter<T>(this IResourceBuilder<T> builder, string name, IEnumerable<string> value)
        where T : AzureBicepResource
    {
        Infrastructure.ValidateIdentifierName(name, nameof(name));
        builder.Resource.Parameters[name] = value;

```

File: sdk/provisioning/Azure.Provisioning/src/Infrastructure.cs (hunk 1 above)
Comment (tg-msft) [reply sdk/provisioning/Azure.Provisioning/src/Infrastructure.cs]: Bicep is underspecified, but a quick test fails for non-ASCII characters.  Tightening that up.

File: sdk/provisioning/Azure.Provisioning/src/Infrastructure.cs (hunk 2 above)
Comment (tg-msft) [reply sdk/provisioning/Azure.Provisioning/src/Infrastructure.cs:181]: Fixed

File: sdk/provisioning/Azure.Provisioning/src/Infrastructure.cs (hunk 3 above)
Comment (tg-msft) [reply sdk/provisioning/Azure.Provisioning/src/Infrastructure.cs]: I don't like adding things to the API that are primarily for my implementation benefit, but it looks like this is a pattern in a lot of the newer validation methods.  Changed.

Comment (eerhardt) [reply sdk/provisioning/Azure.Provisioning/src/Infrastructure.cs]: Note this isn't just for your implementation benefit. We are using this parameterName argument in dotnet/aspire:

https://github.com/dotnet/aspire/blob/a8b1a619d67549243342dfd2c92cfaf26905cb65/src/Aspire.Hosting.Azure/Utils/BicepIdentifierHelpers.cs#L13

**Pull request merged**